File: articles/scientific-computing-for-systems-modeling/r/workflow_assumption_risk.R
# Scientific Computing for Systems Modeling:
# Workflow assumption risk scoring in R.
# Educational example only.

library(tidyverse)

assumptions <- read_csv("../data/workflow_assumptions.csv", show_col_types = FALSE)

scored <- assumptions |>
  mutate(
    risk_score = (1 - confidence) * impact_if_wrong,
    priority = case_when(
      risk_score >= 0.30 ~ "High",
      risk_score >= 0.18 ~ "Medium",
      TRUE ~ "Lower"
    )
  ) |>
  arrange(desc(risk_score))

dir.create("../outputs", showWarnings = FALSE, recursive = TRUE)

write_csv(scored, "../outputs/r_workflow_assumption_risk.csv")

print(scored)

File: articles/scientific-computing-for-systems-modeling/python/workflow_assumption_risk.py
"""
Scientific Computing for Systems Modeling:
Workflow assumption risk scoring.

Educational example only.
"""

from __future__ import annotations

import pandas as pd


def main() -> None:
    assumptions = pd.read_csv("../data/workflow_assumptions.csv")

    assumptions["risk_score"] = (
        (1.0 - assumptions["confidence"]) * assumptions["impact_if_wrong"]
    )

    assumptions = assumptions.sort_values("risk_score", ascending=False)

    print(assumptions)

    assumptions.to_csv("../outputs/python_workflow_assumption_risk.csv", index=False)


if __name__ == "__main__":
    main()
